[Bug]: The experimental ES modules support breaks when using it with some native addons that reexports the native addon module itself, such as isolated-vm

### Version

29.5.0

### Steps to reproduce

1. Clone [`https://github.com/JCMais/jest-issue-native-addons`](https://github.com/JCMais/jest-issue-native-addons)
2. Run: `pnpm/npm install`
3. Run: `pnpm test esBug`
4. Notice the error in the output:
```
❯ pnpm test esBug

> jest-issue-native-addons@1.0.0 test /home/jcm/projects/opensource/jest-issue-native-addons
> NODE_OPTIONS=--experimental-vm-modules pnpm jest --verbose "esBug"

 FAIL  ./esBug.test.ts
  ● Test suite failed to run

    Unexpected closing bracket.
      at @:1:61

      at @:1:61
          at Array.forEach (<anonymous>)

----------|---------|----------|---------|---------|-------------------
File      | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
----------|---------|----------|---------|---------|-------------------
All files |       0 |        0 |       0 |       0 |                   
----------|---------|----------|---------|---------|-------------------
Test Suites: 1 failed, 1 total
Tests:       0 total
Snapshots:   0 total
Time:        0.399 s
Ran all test suites matching /esBug/i.
(node:19378) ExperimentalWarning: VM Modules is an experimental feature and might change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
 ELIFECYCLE  Test failed. See above for more details.
```

In our codebase the error does include the stack trace, which is similar to this one:
```
Unexpected closing brace.
      at @:1:234

      at @:1:234
      at parseSource (node_modules/.pnpm/cjs-module-lexer@1.2.2/node_modules/cjs-module-lexer/lexer.js:185:17)
      at parseCJS (node_modules/.pnpm/cjs-module-lexer@1.2.2/node_modules/cjs-module-lexer/lexer.js:43:5)
          at Array.forEach (<anonymous>)
```


### Expected behavior

It should work and console log the exported function of the isolated-vm module.

### Actual behavior

The error above is thrown and jest crashes.


I am not familiar with Jest's codebase, but I think this could be due to the way reexports are handled here:
https://github.com/jestjs/jest/blob/39f3beda6b396665bebffab94e8d7c45be30454c/packages/jest-runtime/src/index.ts#L833-L850

It looks like it treats the `isolated-vm` export in the `node_modules/isolated-vm/isolated-vm.js` file as a re-export. This is the contents of that file:
```js
module.exports = require('./out/isolated_vm').ivm;
```

It would then try to parse the `./out/isolated_vm` module as a CommonJS module using `cjs-module-lexer`. However, this module is a native addon, which then breaks the lexer cuz it is a binary file.

### Additional context

## Workaround


Change `node_modules/isolated-vm/isolated-vm.js`  from this:
```js
module.exports = require('./out/isolated_vm').ivm;
```
to this:
```js
const ivm = require('./out/isolated_vm').ivm
module.exports = ivm;
```

and it starts working again:
```
❯ pnpm test esBug

> jest-issue-native-addons@1.0.0 test /home/jcm/projects/opensource/jest-issue-native-addons
> NODE_OPTIONS=--experimental-vm-modules pnpm jest --verbose "esBug"

  console.log
    isolated_vm { lib: Lib {} }

      at Object.<anonymous> (esBug.test.ts:6:13)

  console.log
    [Function (anonymous)]

      at Object.<anonymous> (esBug.test.ts:7:13)

(node:13821) ExperimentalWarning: VM Modules is an experimental feature and might change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
 PASS  ./esBug.test.ts
  test
    ✓ works (20 ms)

----------|---------|----------|---------|---------|-------------------
File      | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
----------|---------|----------|---------|---------|-------------------
All files |       0 |        0 |       0 |       0 |                   
----------|---------|----------|---------|---------|-------------------
Test Suites: 1 passed, 1 total
Tests:       1 passed, 1 total
Snapshots:   0 total
Time:        2.533 s
Ran all test suites matching /esBug/i.
```

### Environment

```shell
System:
    OS: Linux 5.10 Ubuntu 22.04.2 LTS 22.04.2 LTS (Jammy Jellyfish)
    CPU: (8) x64 AMD Ryzen 5 5600X 6-Core Processor
  Binaries:
    Node: 18.14.0 - ~/.nvm/versions/node/v18.14.0/bin/node
    Yarn: 1.22.19 - /usr/bin/yarn
    npm: 9.3.1 - ~/.nvm/versions/node/v18.14.0/bin/npm
  npmPackages:
    jest: ^29.5.0 => 29.5.0
```

For me, it's complaining error like this:

```
Cannot find module './out/isolated_vm' from 'node_modules/isolated-vm/isolated-vm.js'

    Require stack:
      node_modules/isolated-vm/isolated-vm.js
      server/util/script.runner.js
      server/router/Rule.js
      server/router/strategry/supportedStrategies.js
      server/router/supportedStrategies.test.js


    However, Jest was able to find:
        'out/isolated_vm.node'

    You might want to include a file extension in your import, or update your 'moduleFileExtensions', which is currently ['js', 'json'].

    See https://jestjs.io/docs/configuration#modulefileextensions-arraystring
```

However, if I add the `'out'` in `moduleFileExtensions`, Jest hangs there forever.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -21,6 +21,7 @@
 - `[jest-config]` Make sure to respect `runInBand` option ([#14578](https://github.com/facebook/jest/pull/14578))
 - `[@jest/expect-utils]` Fix comparison of `DataView` ([#14408](https://github.com/jestjs/jest/pull/14408))
 - `[jest-leak-detector]` Make leak-detector more aggressive when running GC ([#14526](https://github.com/jestjs/jest/pull/14526))
+- `[jest-runtime]` Properly handle re-exported native modules in ESM via CJS ([#14589](https://github.com/jestjs/jest/pull/14589))
 - `[jest-util]` Make sure `isInteractive` works in a browser ([#14552](https://github.com/jestjs/jest/pull/14552))
 - `[pretty-format]` [**BREAKING**] Print `ArrayBuffer` and `DataView` correctly ([#14290](https://github.com/facebook/jest/pull/14290))
 

--- a/e2e/__tests__/nativeEsm.test.ts
+++ b/e2e/__tests__/nativeEsm.test.ts
@@ -97,7 +97,21 @@ test('runs WebAssembly (Wasm) test with native ESM', () => {
 test('does not enforce import assertions', () => {
   const {exitCode, stderr, stdout} = runJest(
     DIR,
-    ['native-esm-missing-import-assertions.test'],
+    ['native-esm-missing-import-assertions.test.js'],
+    {nodeOptions: '--experimental-vm-modules --no-warnings'},
+  );
+
+  const {summary} = extractSummary(stderr);
+
+  expect(summary).toMatchSnapshot();
+  expect(stdout).toBe('');
+  expect(exitCode).toBe(0);
+});
+
+test('properly handle re-exported native modules in ESM via CJS', () => {
+  const {exitCode, stderr, stdout} = runJest(
+    DIR,
+    ['native-esm-native-module.test.js'],
     {nodeOptions: '--experimental-vm-modules --no-warnings'},
   );
 
@@ -113,7 +127,7 @@ onNodeVersions('>=16.12.0', () => {
   test('supports import assertions', () => {
     const {exitCode, stderr, stdout} = runJest(
       DIR,
-      ['native-esm-import-assertions.test'],
+      ['native-esm-import-assertions.test.js'],
       {nodeOptions: '--experimental-vm-modules --no-warnings'},
     );
 
@@ -129,7 +143,7 @@ onNodeVersions('<16.12.0', () => {
   test('syntax error for import assertions', () => {
     const {exitCode, stderr, stdout} = runJest(
       DIR,
-      ['native-esm-import-assertions.test'],
+      ['native-esm-import-assertions.test.js'],
       {nodeOptions: '--experimental-vm-modules --no-warnings'},
     );
 

--- a/e2e/native-esm/__tests__/native-esm-native-module.test.js
+++ b/e2e/native-esm/__tests__/native-esm-native-module.test.js
@@ -1,0 +1,12 @@
+/**
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import ivm from 'isolated-vm';
+
+test('supports native modules exported via CJS', () => {
+  expect(ivm).toBeDefined();
+});

--- a/e2e/native-esm/__tests__/native-esm.test.js
+++ b/e2e/native-esm/__tests__/native-esm.test.js
@@ -47,6 +47,7 @@ test('should support importing node core modules', () => {
     devDependencies: {
       'discord.js': '14.3.0',
       'iso-constants': '^0.1.2',
+      'isolated-vm': '^4.6.0',
       yargs: '^17.5.1',
     },
     jest: {

--- a/e2e/native-esm/package.json
+++ b/e2e/native-esm/package.json
@@ -3,6 +3,7 @@
   "devDependencies": {
     "discord.js": "14.3.0",
     "iso-constants": "^0.1.2",
+    "isolated-vm": "^4.6.0",
     "yargs": "^17.5.1"
   },
   "jest": {

--- a/packages/jest-runtime/src/index.ts
+++ b/packages/jest-runtime/src/index.ts
@@ -850,6 +850,18 @@ export default class Runtime {
       return cachedNamedExports;
     }
 
+    if (path.extname(modulePath) === '.node') {
+      const nativeModule = this.requireModuleOrMock('', modulePath);
+
+      const namedExports = new Set(
+        Object.keys(nativeModule as Record<string, unknown>),
+      );
+
+      this._cjsNamedExports.set(modulePath, namedExports);
+
+      return namedExports;
+    }
+
     const transformedCode =
       this._fileTransforms.get(modulePath)?.code ?? this.readFile(modulePath);
 

